### ISSUE TITLE
unify go-data json and go-wire json

### ISSUE THREAD

**Issue by zramsay** [tendermint/go-amino#22]
@ethanfrey commented on [Thu Mar 02 2017](https://github.com/tendermint/go-data/issues/2)

go-data json makes for nice apis and handles all the types that go-wire does. however, the json format is designed for readability and apis, rather than determinism.  in fact, setting the global `data.Encoder` will definitely lead to different results in different apps. This is fine for an exchange format, but not for anything that goes into app logic (the parsed json as a struct **is** deterministic)

Jae mentioned replacing the go-wire json with the go-data json, which I am not completely against, but we must first check where we rely on the determinism of go-wire json. A quick search in the codebase came up with at least 4 places, where go-wire JSON is used to build sign-bytes.  Which must be deterministic to reliably validate anything. These methods (at least) will need to find an alternative encoding if we modify the go-wire json style.

This is related to the following issue: https://github.com/tendermint/go-wire/issues/8

https://github.com/tendermint/tendermint/blob/master/types/vote.go#L59-L64
https://github.com/tendermint/tendermint/blob/master/types/proposal.go#L43-L48
https://github.com/tendermint/tendermint/blob/master/types/part_set.go#L76-L78
https://github.com/tendermint/tendermint/blob/master/types/block.go#L408-L415

Also see:
https://github.com/tendermint/tendermint/blob/master/types/canonical_json.go

---

@ethanfrey commented on [Fri May 12 2017](https://github.com/tendermint/go-data/issues/2#issuecomment-301092798)

Okay, the rpc interface no longer uses go-wire json in `tendermint:develop`.

We still use this for canonical json generation, which we should revisit as well.

**Comment by ethanfrey** [comment]
Yes, I think we need to clean this up (in tendermint), then can remove the go-wire/json completely.

Check with @ebuchman and @melekes who are working the most on the tendermint codebase.

**Comment by ebuchman** [comment]
See also https://github.com/tendermint/tendermint/issues/608 (note some of that is due to the merkleeyes app I threw together specifically for jepsen testing - could have been more standard)

**Comment by odeke-em** [comment]
Coming here from an internal reference/discussion.
In regards to the first 2 links which still work, the rest have changed as they weren't permalinks, here  are the first two links:
- https://github.com/tendermint/tendermint/blob/7682ad9a60162dd17fd6f61aeed7049a8635ac78/types/vote.go#L60-L65
- https://github.com/tendermint/tendermint/blob/7682ad9a60162dd17fd6f61aeed7049a8635ac78/types/part_set.go#L77-L79

As I had suggested offline, I think we could replace our entire WriteJSON implementation with
```go
func WriteJSON(o interface{}, w io.Writer, n *int, err *error) {
     blob, bErr := json.Marshal(v)
     if bErr != nil {
          *err = bErr
          return
     }
     *n, *err = w.Write(blob)
}
```
and another motivation to move away from our custom JSON serialization implementation is because we don't have support for json.Marshaler interface which the standard library's implementation supports and allows you to implement your custom MarshalJSON method https://golang.org/pkg/encoding/json/#Marshaler
With an exhibit below
```go
package main

import (
    "bytes"
    "encoding/json"
    "fmt"

    "github.com/tendermint/go-wire"
)

type Foo struct {
    A int `json:"a"`
}

func (f *Foo) MarshalJSON() ([]byte, error) {
    return []byte(`{"foo": 10}`), nil
}

func main() {
    var n int
    var err error
    foo := &Foo{A: 10}
    buf := new(bytes.Buffer)
    wire.WriteJSON(foo, buf, &n, &err)
    fmt.Printf("wire: %s\n", buf.Bytes())

    blob, _ := json.Marshal(foo)
    fmt.Printf("stdlib: %s\n", blob)
}
```
which gives
```shell
wire: {"a":10}
stdlib: {"foo":10}
```
showing that our implementation doesn't conform
